MorphicServer/EncryptedString.cs

```diff
@@ -0,0 +1,139 @@
+                }
+                else
+                {
+                    Encrypted = null;
+                }
+            }
+        }
+
+        public class JsonConverter: System.Text.Json.Serialization.JsonConverter<EncryptedString>
+        {
+            public override EncryptedString Read (ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            {
+                var plainText = reader.GetString();
+                var encrypted = new EncryptedString();
+                encrypted.PlainText = plainText;
+                return encrypted;
+            }
+
+            public override void Write (Utf8JsonWriter writer, EncryptedString value, JsonSerializerOptions options)
+            {
+                if (value.PlainText is string plainText)
+                {
+                    writer.WriteStringValue(plainText);
+                }
+                else
+                {
+                    writer.WriteNullValue();
+                }
+            }
+        }
+
+    }
+
+    /// <summary>
+    /// Represents a string that is stored encrypted in the database and only decrypted when needed,
+    /// yet is searchable via a hash representation
+    /// </summary>
+    public class SearchableEncryptedString: EncryptedString
+    {
+
+        public SearchableEncryptedString(string sharedSalt)
+        {
+            SharedSalt = sharedSalt;
+        }
+
+        public HashedData? Hash { get; set; }
+
+        /// <summary>
+        /// Why do we need a shared salt? We need to be able to search
+        /// for the value. If we use random salt for every entry this becomes prohibitively expensive 
+        /// (that being the sole purpose of Salt, after all). This is a trade-off between protecting
+        /// PII and searchability: It's not perfect, but it's sufficient. 
+        /// </summary>
+        [BsonIgnore]
+        public string SharedSalt { get; set; }
+
+        [BsonIgnore]
+        public override string? PlainText{
+            get{
+                return base.PlainText;
+            }
+
+            set{
+                base.PlainText = value;
+                if (value is string plainText)
+                {
+                    Hash = HashedData.FromString(plainText, SharedSalt);
+                }
+                else
+                {
+                    Hash = null;
+                }
+            }
+        }
+
+        public new class JsonConverter: System.Text.Json.Serialization.JsonConverter<SearchableEncryptedString>
+        {
+            public override SearchableEncryptedString Read (ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            {
+                var plainText = reader.GetString();
+                // FIXME: where does the shared salt come from?
+                // Since we're not using a random salt, would simply
+                // using the plainText as the salt achieve the same level of security?
```

Comment:
Is this only used for json-deserialization? If it's a problem finding the configured setting for this field, then the next choice is to simply use NO salt (using a hard-coded salt is a little better, but not enough to justify huge amounts of work to find it in the code). I.e. we could get rid of the hardcoded salt across the board, if this section is an issue.

Reply:
Oh you explained it in the description. It's what I thought, so we can just use 'no salt' instead (or perhaps the string itself), but perhaps we can just ditch the hash here anyway (see my other comment)?